### ISSUE TITLE
Allow overriding the C# namespace used when rendering code

### ISSUE THREAD

**Issue by adamreeve** [gircore/gir.core#1090]
In the Apache Arrow project the GObject libraries have namespaces like "Arrow" or "ArrowDataset", but the existing C# libraries all have a namespace like "Apache.Arrow" or "Apache.Arrow.Dataset".

It would be useful to be able to define a different C# namespace to use when rendering the C# code so that this convention can be kept.

I'm not sure how best to handle this though. Maybe we could support an extra attribute in the gir Namespace element, eg. something like:
```xml
<namespace name="Arrow" csharp:namespace="Apache.Arrow" version="1.0">
  ...
</namespace>
```

Or maybe a new command line argument could be added to support overriding namespaces and avoid having to edit or transform the gir files? Eg. something like `--namespace-overrides=Arrow:Apache.Arrow,ArrowDataset:Apache.Arrow.Dataset`

**Comment by badcel** [comment]
I'm not sure about this one. Do you think it is a must have?

In the beginning of the project I thought about this kind of modifications and I came to the conclusion to avoid customization of the generated code as much as possible.

One thought was to have no way to influence the generated code. The generator should "just work" without any special treatment to the call of `GirTool` (so no parameters + no config files).

Otherwise I could have added "GirCore" to each namespace for example.

I think it makes sense for simplicity reasons and to be as close to the native code as possible to have a strong correlation between the native docs and the binding. This should make life easier for users in the end.

I see that the namespace is not playing an important role in regard to this thoughts so there could be an exception.

On the other hand someone decided for the native libraries to not include the Apache brand name in the namespace why should the C# binding do this? Additionally not having "Apache" in the namespace should be no big deal, too. For me it is more like a marketing decision.

Long story short: I would prefer to not start to "configure" how the code is generated to avoid users to demand more configuration possibilities and keep things simple.

What do you think?

**Comment by adamreeve** [comment]
I'm not sure whether this is a must have, and I get your point about trying to avoid needing customisation parameters as much as possible.

I think it's quite common to have different namespace conventions for different languages though, and being able to support generating different namespaces for different languages from the same GObject repository definition seems quite natural to me. Eg. if you were generating Python code you'd want to have lowercase namespaces like `apache.arrow` to fit the conventions of that language (although they actually use a `pyarrow` namespace to avoid clashing with the `arrow` Python package which is something completely different, and another reason to allow customising the namespace), and for Java you probably want something like `org.apache.arrow`.

This seems similar to the problem of generating code for different languages from protobuf and flatbuffer definitions, and both of those tools allow customising namespace in various ways depending on the platform, eg. see the `csharp_namespace` option for [protobuf](https://protobuf.dev/getting-started/csharptutorial/).

One other reason for wanting this is that I'd been planning on having a separate high level `Apache.Arrow.Dataset` library, and generating the bindings into a namespace like `Apache.Arrow.Dataset.GObj`. I guess by sticking with the `Arrow` namespace I can avoid that problem anyway though.

If you feel strongly against this I'm happy to shelve this idea for now and I can probably work around it.

**Comment by badcel** [comment]
There are already several things done by the generator to match to the language / compiler e.g.:
- The casing of symbols 
- Symbol which are identical to a C# keyword

Those are changes which are done automatically.

For example it would be idiomatic to prefix all interfaces with an "I" but I don't. Or I choose not to implement `INotifyPropertyChanged` because there is a native solution.

I really don't want to create some product which hides that you are actually working with another eco system of libraries. There are so many problems which occur with certain APIs they really can't all be fixed / hidden. The eco systems are to different from a naming perspective alone.

From my point of view it is better to have some awareness on the user side that there is some other eco system in play here.

So my suggestion is to leave it like it is and see if there arises some need from users to actually configure the generated output and then we can bring this topic up again until then let's keep things as simple as possible.

**Comment by adamreeve** [comment]
Ok thanks, I'll close this for now
